### PR TITLE
feat(dapp-connector): Create dApp connector package

### PR DESCRIPTION
**Pull request by michalszorad**
Resolves https://emurgo.atlassian.net/browse/YOMO-1075
- [x] Create new package
- [x] Coverage 100%
- [x] Feature flag
- [x] Readme
- [x] babel-plugin-show-source used only where needed
- [x] Refactor
- [x] Demo
- [x] Comments regarding new tickets

The implementation slightly deviates from [the original proposal](https://emurgo.atlassian.net/wiki/spaces/YOROI/pages/1050476751/DApp+Connector+Technical+Documentation) due to some shortcomings in the original proposal, such as `listAllConnections` which can not be called without initializing `DappConnector` with storage. Once this PR is approved and merged, I'll update the technical specification to match the implementation.

https://github.com/Emurgo/yoroi/assets/19684576/04820100-c906-4eb2-8c2b-1ba1066509bc

### CODE REVIEW

File: apps/wallet-mobile/babel.config.js [hunk 1]
```diff
@@ -3,6 +3,15 @@ module.exports = (api) => {
 
   const presets = ['module:metro-react-native-babel-preset']
   const plugins = [
+    [
+      // Used by @yoroi/dapp-connector to create injectable JS code
+      'babel-plugin-show-source',
```
Comment (michalszorad) [inline apps/wallet-mobile/babel.config.js:8]: Haven't found any other way to stringify a function. Without the plugin `fn.toString()` returns `function() { [bytecode] }`

File: apps/wallet-mobile/ios/Podfile.lock [hunk 2]
```diff
@@ -342,7 +342,8 @@ PODS:
     - React-Core
   - react-native-view-shot (3.8.0):
     - React-Core
-  - react-native-webview (11.26.1):
+  - react-native-webview (13.8.1):
```
Comment (michalszorad) [inline apps/wallet-mobile/ios/Podfile.lock:347]: Updated `react-native-webview` to the latest version

File: apps/wallet-mobile/src/WalletNavigator.tsx [hunk 3]
```diff
@@ -131,6 +137,23 @@ const WalletTabNavigator = () => {
           />
         )}
 
+        {isHaskellShelley(wallet.walletImplementationId) && CONFIG.DAPP_EXPLORER_ENABLED && (
```
Comment (michalszorad) [inline apps/wallet-mobile/src/WalletNavigator.tsx]: dApp explorer menu item shows only if feature flag is enabled

File: apps/wallet-mobile/src/WalletNavigator.tsx [hunk 4]
```diff
@@ -114,7 +120,7 @@ const WalletTabNavigator = () => {
           )}
         </Tab.Screen>
 
-        {isHaskellShelley(wallet.walletImplementationId) && (
+        {isHaskellShelley(wallet.walletImplementationId) && !CONFIG.DAPP_EXPLORER_ENABLED && (
```
Comment (michalszorad) [inline apps/wallet-mobile/src/WalletNavigator.tsx]: Staking is hidden when dApp explorer feature flag is enabled. There is https://emurgo.atlassian.net/browse/YOMO-1238 to refactor the navigation and remove staking from bottom menu

File: apps/wallet-mobile/src/components/SafeArea.tsx [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import React from 'react'
+import {StyleSheet} from 'react-native'
+import {type SafeAreaViewProps, SafeAreaView} from 'react-native-safe-area-context'
+
+import {COLORS} from '../theme'
+
+export const SafeArea = ({children, ...rest}: SafeAreaViewProps) => {
```
Comment (michalszorad) [inline apps/wallet-mobile/src/components/SafeArea.tsx:6]: Moved this to `components` folder as it's used by Governance, and dApp Explorer

File: apps/wallet-mobile/src/features/DappExplorer/common/hooks.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+  const handleWebViewEvent = async (e: WebViewMessageEvent) => {
+    const {data} = e.nativeEvent
+    const webViewUrl = e.nativeEvent.url
+
+    const handlerWallet = {
+      id: MOCK_WALLET_ID,
+      networkId: wallet.networkId,
+      confirmConnection: async (origin: string) => {
+        return new Promise<boolean>((resolve) => {
+          Alert.alert('Confirm connection', `Do you want to connect to ${origin}?`, [
```
Comment (michalszorad) [inline apps/wallet-mobile/src/features/DappExplorer/common/hooks.ts]: The modal will be integrated later in https://emurgo.atlassian.net/browse/YOMO-1236. This is just to showcase the functionality to accept/decline the connection.

File: apps/wallet-mobile/src/features/DappExplorer/common/icons.tsx [hunk 7]
```diff
@@ -0,0 +1,51 @@
+import * as React from 'react'
+import Svg, {Path} from 'react-native-svg'
+
+export const NavigationArrowRightIcon = ({color = '#000000'}) => {
```
Comment (michalszorad) [inline apps/wallet-mobile/src/features/DappExplorer/common/icons.tsx]: Browser specific icons, will be later revised and implemented properly by VAR

File: apps/wallet-mobile/src/features/DappExplorer/common/wallet-config.ts [hunk 8]
```diff
@@ -0,0 +1,4 @@
...
+
+export const WALLET_CONFIG = {name: 'yoroi', apiVersion: '0.3.0', iconUrl}
```
Comment (michalszorad) [inline apps/wallet-mobile/src/features/DappExplorer/common/wallet-config.ts]: What should be the `apiVersion`?

File: apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+  NavigationRefreshIcon,
+  NavigationShareIcon,
+} from '../common/icons'
+
+const DAPP_URL = 'https://app.dexhunter.io/'
+
+const ENABLED_BUTTON_COLOR = '#383E54'
+const DISABLED_BUTTON_COLOR = '#8A92A3'
+
+export const WebBrowserScreen = () => {
```
Comment (michalszorad) [inline apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx]: Full web browser will be implemented in a different PR

File: packages/dapp-connector/src/connector.js [hunk 10]
```diff
@@ -0,0 +1,140 @@
+const initWallet = ({iconUrl, apiVersion, walletName, supportedExtensions, sessionId}) => {
```
Comment (michalszorad) [inline packages/dapp-connector/src/connector.js:17]: This will be injected into web browser.

File: packages/dapp-connector/src/resolver.ts [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+  },
+  isEnabled: async (_params: unknown, context: Context) => {
+    assertOriginsMatch(context)
+    return hasWalletAcceptedConnection(context)
+  },
+  api: {
+    getBalance: async (_params: unknown, context: Context) => {
+      assertOriginsMatch(context)
+      await assertWalletAcceptedConnection(context)
+      return (mockedData as any)[context.wallet.id]?.balance
```
Comment (michalszorad) [inline packages/dapp-connector/src/resolver.ts]: Api method will be implemented here https://emurgo.atlassian.net/browse/YOMO-1191

File: packages/dapp-connector/src/resolver.ts [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+    getNetworkId: ResolvableMethod<number>
+    getRewardAddresses: ResolvableMethod<string[]>
+    getUsedAddresses: ResolvableMethod<string[]>
+  }
+}
+
+export const resolver: Resolver = {
+  logMessage: async (params) => {
+    if (isRecord(params) && isKeyOf('args', params) && Array.isArray(params.args)) {
+      console.log('Log From Dapp Connector:', ...params.args)
```
Comment (michalszorad) [inline packages/dapp-connector/src/resolver.ts:30]: Just for debug purposes

File: apps/wallet-mobile/src/features/DappExplorer/common/hooks.ts [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+const getInitScript = (sessionId: string, dappConnector: DappConnector) => {
+  return dappConnector.getWalletConnectorScript({
+    iconUrl: WALLET_CONFIG.iconUrl,
+    apiVersion: WALLET_CONFIG.apiVersion,
+    walletName: WALLET_CONFIG.name,
+    sessionId,
+  })
+}
+
+const MOCK_WALLET_ID = 'b5d94758-26c5-48b0-af2b-6e68c3ef2dbf'
```
Comment (michalszorad) [inline apps/wallet-mobile/src/features/DappExplorer/common/hooks.ts]: Using a mock wallet until CIP-30 https://emurgo.atlassian.net/browse/YOMO-1191 is not resolved

File: apps/wallet-mobile/src/features/DappExplorer/common/hooks.ts [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+import {Alert} from 'react-native'
+import {WebView, WebViewMessageEvent} from 'react-native-webview'
+
+import {YoroiWallet} from '../../../yoroi-wallets/cardano/types'
+import {Logger} from '../../../yoroi-wallets/logging'
+import {WALLET_CONFIG} from './wallet-config'
+
+const createDappConnector = (appStorage: App.Storage) => {
+  const storage = connectionStorageMaker(appStorage.join('dapp-connections/'))
+  return dappConnectorMaker(storage)
```
Comment (michalszorad) [inline apps/wallet-mobile/src/features/DappExplorer/common/hooks.ts]: May need CSL and other params in the future.

File: apps/wallet-mobile/src/components/SafeArea.tsx [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+      {children}
+    </SafeAreaView>
+  )
+}
+const safeAreaEdges = ['bottom', 'left', 'right', 'bottom'] as const
+
+const styles = StyleSheet.create({
+  root: {
+    flex: 1,
+    backgroundColor: COLORS.WHITE,
```
Comment (banklesss) [inline apps/wallet-mobile/src/components/SafeArea.tsx]: could you implement `theme` here?

Comment (michalszorad) [reply apps/wallet-mobile/src/components/SafeArea.tsx]: Updated

File: apps/wallet-mobile/src/components/SafeArea.tsx (hunk 5 above)
Comment (stackchain) [reply apps/wallet-mobile/src/components/SafeArea.tsx:6]: do you believe this component is actually required? why not only add the default SafeAreaView as most of the screens? It seems to me an overkill.

File: apps/wallet-mobile/src/features/DappExplorer/common/wallet-config.ts (hunk 8 above)
Comment (stackchain) [reply apps/wallet-mobile/src/features/DappExplorer/common/wallet-config.ts]: I dunno, please check with @vsubhuman what is provided on ext, let's use the same if we are providing the same functionalities, maybe is part of CIP30

Comment (stackchain) [inline apps/wallet-mobile/src/features/DappExplorer/common/wallet-config.ts]:
```suggestion
export const walletConfig = {name: 'yoroi', apiVersion: '0.3.0', iconUrl} as const
```

File: apps/wallet-mobile/src/features/DappExplorer/common/wallet-config.ts
Comment (stackchain) [inline apps/wallet-mobile/src/features/DappExplorer/common/wallet-config.ts:1]: rename to constants

File: apps/wallet-mobile/src/features/DappExplorer/common/navigation.ts [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+
+export const NavigationStack = createStackNavigator<Routes>()
+
+export const useNavigateTo = () => {
+  const navigation = useNavigation<StackNavigationProp<Routes>>()
+  const ref = useRef({
+    home: () => navigation.navigate('dapp-explorer-home'),
+    webBrowser: () => navigation.navigate('dapp-explorer-web-browser'),
+  })
+  return ref.current
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/DappExplorer/common/navigation.ts]:
```suggestion
  return React.useRef({
    home: () => navigation.navigate('dapp-explorer-home'),
    webBrowser: () => navigation.navigate('dapp-explorer-web-browser'),
  } as const).current
```

File: apps/wallet-mobile/src/features/DappExplorer/common/navigation.ts [hunk 17]
```diff
@@ -0,0 +1,19 @@
+import {useNavigation} from '@react-navigation/native'
+import {createStackNavigator, StackNavigationProp} from '@react-navigation/stack'
+import {useRef} from 'react'
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/DappExplorer/common/navigation.ts]:
```suggestion
import * as React from 'react'
```

File: apps/wallet-mobile/src/features/DappExplorer/useCases/HomeScreen.tsx [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+
+import {Button} from '../../../components'
+import {useNavigateTo} from '../common/navigation'
+
+export const HomeScreen = () => {
+  const navigateTo = useNavigateTo()
+  return (
+    <View style={styles.root}>
+      <Button
+        title="Go to Web Browser"
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/DappExplorer/useCases/HomeScreen.tsx]: do we have the copies?

File: apps/wallet-mobile/src/features/DappExplorer/useCases/HomeScreen.tsx
Comment (stackchain) [inline apps/wallet-mobile/src/features/DappExplorer/useCases/HomeScreen.tsx:1]: useCases must start with a verb, BrowseDapps, StartTx...

File: apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx
Comment (stackchain) [inline apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx:1]: OperateDapp

File: apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+import {
+  NavigationArrowLeftIcon,
+  NavigationArrowRightIcon,
+  NavigationRefreshIcon,
+  NavigationShareIcon,
+} from '../common/icons'
+
+const DAPP_URL = 'https://app.dexhunter.io/'
+
+const ENABLED_BUTTON_COLOR = '#383E54'
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx]: let's ask @SorinC6 to check theme after merged

File: apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx [hunk 20]
```diff
@@ -0,0 +1,157 @@
+import {useFocusEffect} from '@react-navigation/native'
+import React, {useState} from 'react'
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx]:
```suggestion
import * as React from 'react'
```

File: apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+    justifyContent: 'center',
+  },
+})
+
+function TabScreen({sessionId}: {sessionId: string}) {
+  const ref = React.useRef<WebView | null>(null)
+  const wallet = useSelectedWallet()
+  const {handleEvent, initScript} = useConnectWalletToWebView(wallet, ref, sessionId)
+  const [currentUrl, setCurrentUrl] = useState(DAPP_URL)
+  const [inputUrl, setInputUrl] = useState('')
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx]:
```suggestion
  const [currentUrl, setCurrentUrl] = React.useState(DAPP_URL)
  const [inputUrl, setInputUrl] = React.useState('')
```

File: apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+
+  const [canGoBack, setCanGoBack] = React.useState(false)
+  const [canGoForward, setCanGoForward] = React.useState(false)
+
+  const [isFocused, setIsFocused] = useState(false)
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx]:
```suggestion
  const [isFocused, setIsFocused] = React.useState(false)
```

File: apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+
+  useFocusEffect(() => {
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx]: focus effect recommendation is to wrap within a callback

File: apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx [hunk 24]
```diff
@@ -0,0 +1,157 @@
...
+    const normalizedUrl = url.startsWith('http') ? url : `https://${url}`
+    setCurrentUrl(normalizedUrl)
+  }
+
+  const handleLoadingEnd = (url: string) => {
+    setInputUrl(url)
+    setCurrentUrl(url)
+    ref.current?.injectJavaScript(initScript)
+  }
+
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx]:
```suggestion
  const handleOnReload = () => {
    ref.current?.reload()
  }

  const handleOnNavigateBack = () => {
    ref.current?.goBack()
  }

  const handleOnNavigateForward = () => {
    ref.current?.goForward()
  }

  const handleOnShare = () => {
    if (Platform.OS === 'android') {
      Share.share({message: currentUrl})
    } else {
      Share.share({url: currentUrl})
    }
  }

  const handleOnSubmitUrl = (url: string) => {
    const normalizedUrl = url.startsWith('http') ? url : `https://${url}`
    setCurrentUrl(normalizedUrl)
  }

  const handleOnLoaded = (url: string) => {
    setInputUrl(url)
    setCurrentUrl(url)
    ref.current?.injectJavaScript(initScript)
  }

```

File: apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx [hunk 25]
```diff
@@ -0,0 +1,157 @@
...
+  navigationButton: {
+    flex: 1,
+    width: 24,
+    height: 24,
+    alignItems: 'center',
+    justifyContent: 'center',
+  },
+})
+
+function TabScreen({sessionId}: {sessionId: string}) {
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx]: use Screen only for Modal first child and the component that takes over the screen.

File: apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx [hunk 26]
```diff
@@ -0,0 +1,157 @@
...
+  }
+
+  const handleLoadingEnd = (url: string) => {
+    setInputUrl(url)
+    setCurrentUrl(url)
+    ref.current?.injectJavaScript(initScript)
+  }
+
+  return (
+    <View style={{flex: 1}}>
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx]: if it's a

Comment (stackchain) [inline apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx]: if it's a screen it must be wrapped within the SafeAreaView, here, or at the Navigator.

File: packages/dapp-connector/src/connector.js [hunk 27]
```diff
@@ -0,0 +1,140 @@
...
+      promise.reject(handleError(error))
+    } else {
+      promise.resolve(result)
+    }
+  })
+
+  const createApi = (cardanoEnableResponse) => {
+    if (!cardanoEnableResponse) {
+      logMessage('User Rejected')
+      throw {code: -3, info: 'User Rejected'}
```
Comment (stackchain) [inline packages/dapp-connector/src/connector.js]: are these errors standardize?

File: packages/dapp-connector/src/connector.js [hunk 28]
```diff
@@ -0,0 +1,140 @@
...
+
+  const serializeError = (error) => {
+    if (error instanceof Error) {
+      return error.name + ': ' + error.message
+    }
+    return JSON.stringify(error)
+  }
+
+  window.addEventListener('message', (event) => {
+    logMessage('Received message' + JSON.stringify(event.data))
```
Comment (stackchain) [inline packages/dapp-connector/src/connector.js:103]: can we prefix all logs with `[yoroi-mobile-connector]: `?

File: packages/dapp-connector/src/storage.ts
Comment (stackchain) [inline packages/dapp-connector/src/storage.ts:1]: `src/adapters/async-storage.ts`

File: packages/dapp-connector/src/storage.ts [hunk 29]
```diff
@@ -0,0 +1,37 @@
+import {App} from '@yoroi/types'
+
+export const connectionStorageMaker = (storage: App.Storage): Storage => {
+  const listAllConnections = async () => (await storage.getItem<DappConnection[]>('connections')) || []
+
+  const removeConnection = async (connection: DappConnection) => {
```
Comment (stackchain) [inline packages/dapp-connector/src/storage.ts]: use the correct convention
```{
 connections: 
  {
    save..,
    read..,
    yadayada
  }
}
```

File: packages/dapp-connector/src/storage.ts [hunk 30]
```diff
@@ -0,0 +1,37 @@
...
+  }
+
+  const addConnection = async (connection: DappConnection) => {
+    const connections = await listAllConnections()
+    const newConnections = [...connections, connection]
+    await storage.setItem('connections', newConnections)
+  }
+
+  return {
+    listAllConnections,
```
Comment (stackchain) [inline packages/dapp-connector/src/storage.ts]: storage read/save/delete/key, ==> list is manager responsibility

Comment (stackchain) [reply packages/dapp-connector/src/storage.ts]: on manager 
```{
   connections: {
     list: {
       byName....
       yadada
     }
   }
}
```

File: apps/wallet-mobile/src/features/DappExplorer/useCases/HomeScreen.tsx (hunk 18 above)
Comment (michalszorad) [reply apps/wallet-mobile/src/features/DappExplorer/useCases/HomeScreen.tsx]: This is just a placeholder, as the UI will be implemented by VAR

File: packages/dapp-connector/src/storage.ts [hunk 31]
```diff
@@ -0,0 +1,37 @@
...
+    removeConnection,
+    addConnection,
+  }
+}
+
+const areConnectionsEqual = (a: DappConnection, b: DappConnection) =>
+  a.walletId === b.walletId && a.dappOrigin === b.dappOrigin
+
+export interface DappConnection {
```
Comment (stackchain) [inline packages/dapp-connector/src/storage.ts]: for now export this to a local `types.ts` later it should be lifted to `@yoroi/types`

File: packages/dapp-connector/src/storage.ts [hunk 32]
```diff
@@ -0,0 +1,37 @@
...
+  }
+
+  return {
+    listAllConnections,
+    removeConnection,
+    addConnection,
+  }
+}
+
+const areConnectionsEqual = (a: DappConnection, b: DappConnection) =>
```
Comment (stackchain) [inline packages/dapp-connector/src/storage.ts]: ~is it in use?~ maybe it shouldn't store same connections

File: packages/dapp-connector/src/storage.ts [hunk 33]
```diff
@@ -0,0 +1,37 @@
...
+  a.walletId === b.walletId && a.dappOrigin === b.dappOrigin
+
+export interface DappConnection {
+  walletId: string
+  dappOrigin: string
+}
+
+export interface Storage {
```
Comment (stackchain) [inline packages/dapp-connector/src/storage.ts]: will be lifted -> yoroi/types

File: packages/dapp-connector/src/connector.js [hunk 34]
```diff
@@ -0,0 +1,140 @@
...
+  if (window.cardano && window.cardano[walletName]) return
+
+  const postMessage = (data) => {
+    if (window.ReactNativeWebView) {
+      window.ReactNativeWebView.postMessage(JSON.stringify(data))
+      return
+    }
+    window.postMessage(JSON.stringify(data))
+  }
+  const promisesMap = new Map()
```
Comment (stackchain) [inline packages/dapp-connector/src/connector.js:47]: looks like a great candidate for WeakMap.

File: packages/dapp-connector/src/connector.js [hunk 35]
```diff
@@ -0,0 +1,140 @@
...
+  const walletObj = Object.freeze({
+    enable,
+    isEnabled,
+    icon: iconUrl,
+    apiVersion: apiVersion,
+    name: walletName,
+    supportedExtensions: Object.freeze(supportedExtensions),
+  })
+
+  window.cardano = window.cardano || {}
```
Comment (stackchain) [inline packages/dapp-connector/src/connector.js]: prefer ?? for defaulting
```suggestion
  window.cardano = window.cardano ?? {}
```

File: packages/dapp-connector/src/connector.js (hunk 34 above)
Comment (stackchain) [reply packages/dapp-connector/src/connector.js:47]: is the order relevant?

File: packages/dapp-connector/src/connector.js [hunk 36]
```diff
@@ -0,0 +1,140 @@
...
+
+    enabling = false
+    return {
+      getBalance: (...args) => callExternalMethod('api.getBalance', args),
+      getChangeAddress: (...args) => callExternalMethod('api.getChangeAddress', args),
+      getNetworkId: (...args) => callExternalMethod('api.getNetworkId', args),
+      getRewardAddresses: (...args) => callExternalMethod('api.getRewardAddresses', args),
+      getUsedAddresses: (...args) => callExternalMethod('api.getUsedAddresses', args),
+    }
+  }
```
Comment (stackchain) [inline packages/dapp-connector/src/connector.js:134]: @michaeljscript can we add JS Docs
/**
 * @typedef {Object} ExternalAPI
 * @property {(...args: any[]) => Promise<any>} getBalance Function to get balance.
 * @property {(...args: any[]) => Promise<any>} getChangeAddress Function to get change address.
 * @property {(...args: any[]) => Promise<any>} getNetworkId Function to get network ID.
 * @property {(...args: any[]) => Promise<any>} getRewardAddresses Function to get reward addresses.
 * @property {(...args: any[]) => Promise<any>} getUsedAddresses Function to get used addresses.
 */

/**
 * @type {ExternalAPI}
 */

This will make any IDE, to know about the types.

File: packages/dapp-connector/src/dapp-connector.test.ts [hunk 37]
```diff
@@ -0,0 +1,277 @@
+import {dappConnectorMaker, connectionStorageMaker} from './index'
```
Comment (stackchain) [inline packages/dapp-connector/src/dapp-connector.test.ts]: avoid importing from index (circular dependencies)

File: packages/dapp-connector/src/resolver.ts (hunk 11 above)
Comment (stackchain) [reply packages/dapp-connector/src/resolver.ts]: CBOR on web, genius idea, ok, 2 things, it needs checking with backend, we can't serialize back, so we need to get it straight from utxo module.

File: apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx (hunk 26 above)
Comment (michalszorad) [reply apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx]: Safe area is in the dapp explorer navigator, should I move it to individual screen files?

File: packages/dapp-connector/src/connector.js (hunk 27 above)
Comment (michalszorad) [reply packages/dapp-connector/src/connector.js]: Yes, it's a part of CIP-30

File: packages/dapp-connector/src/connector.js (hunk 34 above)
Comment (michalszorad) [reply packages/dapp-connector/src/connector.js:47]: Order is not relevant, we use IDs to identify promises

File: packages/dapp-connector/src/connector.js (hunk 27 above)
Comment (stackchain) [reply packages/dapp-connector/src/connector.js]: should it be part of the message or should we throw as an Object? --> throw new Error(....), or CIP30Error extends Error, and in this class we have code + info ?, is it ok throwing an Object?

File: apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx (hunk 26 above)
Comment (stackchain) [reply apps/wallet-mobile/src/features/DappExplorer/useCases/WebBrowserScreen.tsx]: na, if it is on nav better, we just can't navigate the user off, otherwise it will wrap twice the edges.

File: packages/dapp-connector/src/connector.js (hunk 27 above)
Comment (michalszorad) [reply packages/dapp-connector/src/connector.js]: CIP30Error sounds good I'll add that. Throwing an object is definitely not ideal, but that's what I believe we have on extension and did not want to deviate too much with the implementation.

File: apps/wallet-mobile/src/features/DappExplorer/common/wallet-config.ts (hunk 8 above)
Comment (michalszorad) [reply apps/wallet-mobile/src/features/DappExplorer/common/wallet-config.ts]: Should be set to `1` that's now updated

File: apps/wallet-mobile/src/components/SafeArea.tsx (hunk 5 above)
Comment (michalszorad) [reply apps/wallet-mobile/src/components/SafeArea.tsx:6]: I've reused it on quite many places in storybook, but happy to remove if that's the preference

File: packages/dapp-connector/src/dapp-connector.test.ts (hunk 37 above)
Comment (michalszorad) [reply packages/dapp-connector/src/dapp-connector.test.ts]: Updated

File: packages/dapp-connector/src/connector.js (hunk 36 above)
Comment (michalszorad) [reply packages/dapp-connector/src/connector.js:134]: Updated

File: packages/dapp-connector/src/connector.js (hunk 27 above)
Comment (michalszorad) [reply packages/dapp-connector/src/connector.js]: Updated

File: apps/wallet-mobile/src/features/Discover/BrowserNavigator.tsx [hunk 38]
```diff
@@ -13,22 +18,45 @@ const Tab = createStackNavigator<BrowserRoutes>()
...
   )
 }
 
+const createDappConnector = (appStorage: App.Storage, wallet: YoroiWallet) => {
+  const handlerWallet = {
+    id: wallet.id,
+    networkId: wallet.networkId,
+    confirmConnection: async (origin: string) => {
+      return new Promise<boolean>((resolve) => {
+        Alert.alert('Confirm connection', `Do you want to connect to ${origin}?`, [
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/Discover/BrowserNavigator.tsx:50]: @michaeljscript to leverage translations here, whether this become a hook itself, or pass down strings.

File: packages/dapp-connector/src/dapp-connector.ts [hunk 39]
```diff
@@ -0,0 +1,54 @@
...
+  }) => {
+    return connectWallet({...props, supportedExtensions})
+  }
+
+  public handleEvent = async (
+    eventData: string,
+    trustedUrl: string,
+    sendMessage: (id: string, result: unknown, error?: Error) => void,
+  ) => {
+    return await handleEvent(eventData, trustedUrl, this.wallet, sendMessage, this.storage)
```
Comment (stackchain) [inline packages/dapp-connector/src/dapp-connector.ts]: resolverHandleEvent, (no-shadowing) - we need to add this rule on packages.

File: packages/dapp-connector/src/resolver.ts [hunk 40]
```diff
@@ -0,0 +1,151 @@
...
+  },
+  isEnabled: async (_params: unknown, context: Context) => {
+    assertOriginsMatch(context)
+    return hasWalletAcceptedConnection(context)
+  },
+  api: {
+    getBalance: async (_params: unknown, context: Context) => {
+      assertOriginsMatch(context)
+      await assertWalletAcceptedConnection(context)
+      return mockedData[mockWalletId1].balance
```
Comment (stackchain) [inline packages/dapp-connector/src/resolver.ts:49]: still mocking cuz we are missing are lacking the CBORs from API?

File: packages/dapp-connector/src/resolver.ts [hunk 41]
```diff
@@ -0,0 +1,151 @@
...
+  eventData: string,
+  trustedUrl: string,
+  wallet: Wallet,
+  sendMessage: (id: string, result: unknown, error?: Error) => void,
+  storage: Storage,
+) => {
+  const trustedOrigin = new URL(trustedUrl).origin
+  const {id, method, params} = JSON.parse(eventData)
+  await handleMethod(method, params, {origin: trustedOrigin, wallet, storage})
+    .then((result) => method !== LOG_MESSAGE_EVENT && sendMessage(id, result))
```
Comment (stackchain) [inline packages/dapp-connector/src/resolver.ts]: I'm ok with thenables, we just need to fix a convention.

**Review by stackchain — APPROVED** [review]
looking solid to me, 2 points, 2 questions

File: packages/dapp-connector/src/resolver.ts (hunk 41 above)
Comment (michalszorad) [reply packages/dapp-connector/src/resolver.ts]: I'll update

File: packages/dapp-connector/src/resolver.ts (hunk 40 above)
Comment (michalszorad) [reply packages/dapp-connector/src/resolver.ts:49]: CIP30 will be implemented in a different PR for easier review

File: apps/wallet-mobile/src/features/Discover/BrowserNavigator.tsx (hunk 38 above)
Comment (michalszorad) [reply apps/wallet-mobile/src/features/Discover/BrowserNavigator.tsx:50]: I've added a todo, as this will use a modal with translations instead of the alert
